domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt

```diff
@@ -116,11 +133,51 @@ class AppStartupStateController @Inject constructor(
   }
 
   private fun computeAppStartupMode(onboardingState: OnboardingState): StartupMode {
-    return when {
-      hasAppExpired() -> StartupMode.APP_IS_DEPRECATED
-      onboardingState.alreadyOnboardedApp -> StartupMode.USER_IS_ONBOARDED
-      else -> StartupMode.USER_NOT_YET_ONBOARDED
+    // Return old logic if app and os feature flag is not enabled
+    if (!enableAppAndOsDeprecation.value) {
+      return when {
+        hasAppExpired() -> StartupMode.APP_IS_DEPRECATED
+        onboardingState.alreadyOnboardedApp -> StartupMode.USER_IS_ONBOARDED
+        else -> StartupMode.USER_NOT_YET_ONBOARDED
+      }
     }
+
+    val deprecationDataProvider = deprecationController.getDeprecationDatabase()
+
+    var deprecationDatabase = DeprecationResponseDatabase.newBuilder().build()
+
+    runBlocking {
+      deprecationDataProvider.retrieveData().transform {
+        deprecationDatabase = it
+      }
+    }
```

Comment:
We typically do not use **runBlocking** in production code. Dataproviders have functions for transforming the data, and you could do this in a suspend function instead. The ultimate choice will depend on the refactor per my suugestion below.

Reply:
A simple controller handling data provider transformations: https://github.com/oppia/oppia-android/blob/develop/domain/src/main/java/org/oppia/android/domain/survey/SurveyGatingController.kt